project_timesheet_holidays_contract/models/hr_employee.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from datetime import datetime
+
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+
+
+class HrEmployee(models.Model):
+    _inherit = "hr.employee"
+
+    def _get_calendar(self, date_from=None):
+        if not date_from:  # no `date_from` => call `super()`
+            return super()._get_calendar(date_from)
+
+        date_from = (
+            date_from if not isinstance(date_from, datetime) else date_from.date()
+        )
+        date_to = self.env.context.get("date_to", None)
+        if date_to is None:
+            date_to = fields.Date.today()
+        date_to = date_to if not isinstance(date_to, datetime) else date_to.date()
+
+        contracts = (
+            self.sudo()
+            .with_context(active_test=True)
+            .env["hr.contract"]
+            .search(
+                [
+                    ("employee_id", "=", self.id),
+                    ("date_start", "<=", date_from),
+                    "|",
+                    ("date_end", ">=", date_to),
+                    ("date_end", "=", False),
+                ]
+            )
+        )
+
+        # If only one contract, return it regardless of its state
+        if len(contracts) == 1:
+            return contracts[0].resource_calendar_id
+
+        # Filter out cancelled contracts if there are multiple contracts
+        valid_contracts = contracts.filtered(lambda c: c.state != "cancel")
```

Comment:
I tested this w/ Odoo standard code, and I think states are not handled correctly.

Contracts can be differentiated in these macro-groups:
- `state == 'close'`: the contract is considered as expired for the employee
- `state == 'cancel'`: the contract is considered as cancelled for the employee
- `state == 'open' and kanban_state != 'blocked'`: the contract is considered as running for the employee
- `state == 'open' and kanban_state == 'blocked'`: the contract is considered as running for the employee, but will be closed as soon as the end date is reached
- `state == 'draft' and kanban_state != 'done'`: the contract is not running for the employee
- `state == 'draft' and kanban_state == 'done'`: the contract is not running for the employee, but will be moved to "open" as soon as the contract's start date is met

This is something that should be taken into account when retrieving the employee's contract, and therefore its calendar.